### PR TITLE
Support PyTorch ET simulation on ASTRA-sim

### PR DESCRIPTION
**Pull request by TaekyungHeo**
Enhance Astra-sim to accommodate Chakra traces converted from a consolidated Execution trace (Pytorch + Kineto).

### CODE REVIEW

File: astra-sim/workload/HardwareResource.cc [hunk 1]
```diff
@@ -12,45 +12,37 @@ using namespace Chakra;
...
   }
 }
 
 bool HardwareResource::is_available(
     const shared_ptr<Chakra::ETFeederNode> node) const {
-  if ((node->getChakraNode()->node_type() == ChakraNodeType::COMP_NODE) &&
-      (num_in_flight_comps >= num_npus)) {
-    return false;
+  if (node->is_cpu_op()) {
+      if (num_in_flight_cpu_ops == 0) {
```
Comment (beckmabd) [inline astra-sim/workload/HardwareResource.cc:36]: So is this assuming there are only one CPU or one GPU available?

File: astra-sim/workload/Workload.cc [hunk 2]
```diff
@@ -151,127 +155,145 @@ void Workload::issue_remote_mem(shared_ptr<Chakra::ETFeederNode> node) {
...
-          node->getChakraNode()->comm_priority());
-      collective_comm_node_id_map[fp->my_id] = node->getChakraNode()->id();
+          node->comm_priority());
+      collective_comm_node_id_map[fp->my_id] = node->id();
+      collective_comm_wrapper_map[fp->my_id] = fp;
       fp->set_notifier(this, EventType::CollectiveCommunicationFinished);
 
     } else if (
-        node->getChakraNode()->comm_type() ==
+        node->comm_type() ==
```
Comment (beckmabd) [inline astra-sim/workload/Workload.cc:213]: What is going on with the coding style here?  I realize you are just following the prior coding style, but I don't understand why there is a new line here.  The first half of the condition would fit on the prior line.  Can you make this and the following "else if" style follow the pior style shown on line 201?

File: astra-sim/workload/Workload.cc [hunk 3]
```diff
@@ -310,6 +332,10 @@ void Workload::call(EventType event, CallData* data) {
 
     et_feeder->removeNode(node_id);
 
+    // The Dataset class provides statistics that should be used later to dump more stats in the workload layer
```
Comment (beckmabd) [inline astra-sim/workload/Workload.cc:335]: Spread this comment across 2 lines.  No line longer than 78 chars.

**Review by beckmabd — CHANGES_REQUESTED** [review]
I have a few questions that I'd like to get your feedback on before I approve it.  Thanks!

File: astra-sim/workload/Workload.cc [hunk 4]
```diff
@@ -151,127 +155,145 @@ void Workload::issue_remote_mem(shared_ptr<Chakra::ETFeederNode> node) {
...
+      // broadcast colelctive has not been implemented in ASTRA-SIM yet.
+      // So, we just use its real system mesurements
+      DataSet* fp = new DataSet(1);
+      fp->set_notifier(this, EventType::CollectiveCommunicationFinished);
+      sys->register_event(
+        fp,
+        EventType::General,
+        nullptr,
+        // chakra runtimes are in microseconds and we should convert it into nanoseconds
+        node->runtime() * 1000);
```
[the remark this answers is not in the document]
Comment (TaekyungHeo) [reply astra-sim/workload/Workload.cc:259]: Sure, I will add an assertion here. Yes, we should assume that all broadcast collective nodes have a runtime.

File: astra-sim/system/Sys.hh [hunk 5]
```diff
@@ -309,6 +309,9 @@ class Sys : public Callable {
 
   // statistics
   bool trace_enabled;
+
+  //only replay the ops and do not simulate any op
```
[the remark this answers is not in the document]
Comment (TaekyungHeo) [reply astra-sim/system/Sys.hh:313]: Sure, I will update the comment. By replaying, ASTRA-sim utilized the measured runtime for all operators. In other words, we are not running any compute models or network models.

File: astra-sim/workload/Workload.cc (hunk 2 above)
Comment (TaekyungHeo) [reply astra-sim/workload/Workload.cc:213]: You will understand this change when you take a look at another PR in Chakra: https://github.com/TaekyungHeo/chakra-1/pull/2. ASTRA-sim relies on et_feeder in Chakra for node definitions and the frontend. We've made updates to how we retrieve comm_type in that PR.

File: astra-sim/workload/Workload.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (TaekyungHeo) [reply astra-sim/workload/Workload.cc:263]: Please find the updated code.

File: astra-sim/system/Sys.hh (hunk 5 above)
[the remark this answers is not in the document]
Comment (TaekyungHeo) [reply astra-sim/system/Sys.hh]: Please find the updated code.

File: astra-sim/system/Sys.hh [hunk 6]
```diff
@@ -309,6 +309,9 @@ class Sys : public Callable {
 
   // statistics
   bool trace_enabled;
+
+  // replay operators based on their runtime and do not run performance models
```
Comment (TaekyungHeo) [inline astra-sim/system/Sys.hh:313]: @beckmabd Updated the comment.

File: astra-sim/workload/HardwareResource.cc [hunk 7]
```diff
@@ -12,45 +12,37 @@ using namespace Chakra;
...
-      (node->getChakraNode()->node_type() == ChakraNodeType::MEM_STORE_NODE)) {
-    ++num_in_flight_mem_reqs;
-  } else if (node->getChakraNode()->node_type() == ChakraNodeType::COMP_NODE) {
-    ++num_in_flight_comps;
-  } else if (
-      (node->getChakraNode()->node_type() == ChakraNodeType::COMM_SEND_NODE) ||
-      (node->getChakraNode()->node_type() == ChakraNodeType::COMM_RECV_NODE) ||
-      (node->getChakraNode()->node_type() == ChakraNodeType::COMM_COLL_NODE)) {
-    ++num_in_flight_comms;
+  if (node->is_cpu_op()) {
```
[the remark this answers is not in the document]
Comment (TaekyungHeo) [reply astra-sim/workload/HardwareResource.cc:18]: * Saeed found out we need to model both a CPU and GPU in ASTRA-sim to run PyTorch execution traces on it. After post-processing, the execution traces will have both CPU and GPU operators. That's why we want to differentiate between CPU ops and GPU ops.
* Correct, resource modeling should happen in the system layer. We didn't get a chance to refactor the codes yet.
* We can implement such a feature, but it's not available now.

File: astra-sim/system/Sys.hh (hunk 6 above)
Comment (srinivas212) [reply astra-sim/system/Sys.hh:313]: Better to say "skip simulation for operator and use current duration". 

Replay means other things given we have chakra/param replay?

File: astra-sim/workload/HardwareResource.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (srinivas212) [reply astra-sim/workload/HardwareResource.cc:18]: Re. NPU - from a modeling perspective we could think of only modeling "npu" in system layer so have cpu instance of it and gpu instance of it? so in effect you have different types of npus. Agree with @TaekyungHeo  we do need to have this distinction from chakra / workload trace.

File: astra-sim/workload/HardwareResource.cc (hunk 1 above)
Comment (srinivas212) [reply astra-sim/workload/HardwareResource.cc:36]: from each chakra trace perspective - you have one trace per GPU and it has corresponding CPU thread.

File: astra-sim/system/Sys.hh (hunk 6 above)
Comment (TaekyungHeo) [reply astra-sim/system/Sys.hh]: @srinivas212 , Updated, thanks!

File: astra-sim/workload/Workload.cc [hunk 8]
```diff
@@ -103,45 +102,53 @@ void Workload::issue_dep_free_nodes() {
...
-           << ",node->name=" << node->getChakraNode()->name() << endl;
-    }
-    issue_remote_mem(node);
-  } else if (node->getChakraNode()->node_type() == ChakraNodeType::COMP_NODE) {
-    if ((node->getChakraNode()->simulated_run_time() == 0) &&
-        (node->getChakraNode()->num_ops() == 0)) {
-      skip_invalid(node);
-    } else {
+  if (sys->replay_only) {
+      issue_comp(node);
```
Comment (beckmabd) [inline astra-sim/workload/Workload.cc:112]: If replay_only means we are relying on the recorded execution times for all nodes, including communication nodes, then it is confusing here to call "issue_comp" on all nodes.  We should rename "issue_comp" to something to the effect "advance_time" or "use_recorded_latency".  Also within that newly renamed function, we should assert that the passed in node has a recorded time.

File: astra-sim/system/Sys.hh [hunk 9]
```diff
@@ -309,6 +309,9 @@ class Sys : public Callable {
 
   // statistics
   bool trace_enabled;
+
+  // skip simulation for operator and use current duration
+  bool replay_only;
```
[the remark this answers is not in the document]
Comment (beckmabd) [reply astra-sim/system/Sys.hh:314]: Please modify the comment to say "skip simulation for all nodes".  Assuming I understand this mode correctly.

File: astra-sim/workload/HardwareResource.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (beckmabd) [reply astra-sim/workload/HardwareResource.cc:18]: Ok, could you please add a comment highlighting this resource modeling should be moved to the system layer.

I'm fine adding the CPU resource code.  In general, I think we should have a positive additute for adding features that help our developer/user base.  In our next meeting, I'd be happy to discuss the details of the cost/benefit burden reviewers need to have for open source projects.  I just wanted to make sure that this code is not requiring CPU-specific layers in the ET.  Based on my visual analysis, I don't think so, but I just wanted to make sure.  I think your response just points out that currently the resource modeling logic allows only one GPU operator and one CPU operator outstanding per NPU at a time, correct?

File: astra-sim/workload/HardwareResource.cc (hunk 1 above)
Comment (beckmabd) [reply astra-sim/workload/HardwareResource.cc:36]: But you don't need a cooresponding CPU thread, correct?  There are many different ways users can create charka traces.  It is not just directly from PyTorch.

File: astra-sim/workload/Workload.cc [hunk 10]
```diff
@@ -151,127 +158,146 @@ void Workload::issue_remote_mem(shared_ptr<Chakra::ETFeederNode> node) {
   WorkloadLayerHandlerData* wlhd = new WorkloadLayerHandlerData;
   wlhd->sys_id = sys->id;
   wlhd->workload = this;
-  wlhd->node_id = node->getChakraNode()->id();
-  sys->remote_mem->issue(node->getChakraNode()->tensor_size(), wlhd);
+  wlhd->node_id = node->id();
+  sys->remote_mem->issue(node->tensor_size(), wlhd);
 }
 
 void Workload::issue_comp(shared_ptr<Chakra::ETFeederNode> node) {
```
Comment (beckmabd) [inline astra-sim/workload/Workload.cc:165]: Note, I think we should rename this function to something more meaningful if I understand these changes correctly.  This function is not just for computation layers.

File: astra-sim/workload/Workload.cc [hunk 11]
```diff
@@ -151,127 +158,146 @@ void Workload::issue_remote_mem(shared_ptr<Chakra::ETFeederNode> node) {
...
-    wlhd->node_id = node->getChakraNode()->id();
+    wlhd->node_id = node->id();
 
     sys->register_event(
         this,
         EventType::General,
         wlhd,
-        node->getChakraNode()->simulated_run_time());
+        // chakra runtimes are in microseconds and we should convert it into nanoseconds
+        node->runtime() * 1000);
```
Comment (beckmabd) [inline astra-sim/workload/Workload.cc:190]: we should add an assert to make sure the node has a valid runtime.

File: astra-sim/workload/Workload.cc (hunk 2 above)
Comment (beckmabd) [reply astra-sim/workload/Workload.cc:216]: I hope to get the time to review your other changes soon, however my suggestion was much simplier and just related to the coding style.  I simply wanted you to change:

"    } else if (
        node->comm_type() =="

to

"    } else if (node->comm_type() == "

That matches the style in the rest of the file.

File: astra-sim/workload/Workload.cc [hunk 12]
```diff
@@ -310,6 +336,11 @@ void Workload::call(EventType event, CallData* data) {
 
     et_feeder->removeNode(node_id);
 
+    // The Dataset class provides statistics that should be used later to dump
+    // more statistics in the workload layer
```
[the remark this answers is not in the document]
Comment (beckmabd) [reply astra-sim/workload/Workload.cc:340]: Thanks!

**Review by beckmabd — CHANGES_REQUESTED** [review]
Thanks for the feedback.  I now understand this code a lot better.  Just a few more suggestions so that the assumptions behind these changes are clear.

File: astra-sim/workload/HardwareResource.cc [hunk 13]
```diff
@@ -3,6 +3,8 @@ This source code is licensed under the MIT license found in the
 LICENSE file in the root directory of this source tree.
 *******************************************************************************/
 
+// TODO: HardwareResource.cc should be moved to the system layer.
```
Comment (TaekyungHeo) [inline astra-sim/workload/HardwareResource.cc:6]: @beckmabd

File: astra-sim/workload/HardwareResource.hh [hunk 14]
```diff
@@ -3,6 +3,8 @@ This source code is licensed under the MIT license found in the
 LICENSE file in the root directory of this source tree.
 *******************************************************************************/
 
+// TODO: HardwareResource.hh should be moved to the system layer.
```
Comment (TaekyungHeo) [inline astra-sim/workload/HardwareResource.hh:6]: @beckmabd

File: astra-sim/workload/Workload.cc [hunk 15]
```diff
@@ -103,175 +102,201 @@ void Workload::issue_dep_free_nodes() {
...
+      skip_invalid(node);
     }
-    issue_comm(node);
-  } else if (
-      node->getChakraNode()->node_type() == ChakraNodeType::INVALID_NODE) {
-    skip_invalid(node);
   }
 }
 
+void Workload::issue_replay(shared_ptr<Chakra::ETFeederNode> node) {
```
Comment (TaekyungHeo) [inline astra-sim/workload/Workload.cc:155]: @beckmabd

File: astra-sim/system/Sys.hh (hunk 9 above)
[the remark this answers is not in the document]
Comment (TaekyungHeo) [reply astra-sim/system/Sys.hh:314]: Updated!

File: astra-sim/workload/HardwareResource.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (TaekyungHeo) [reply astra-sim/workload/HardwareResource.cc:20]: Updated! Yes, currently only one outstanding op is allowed for CPU op and GPU op, respectively.

File: astra-sim/workload/HardwareResource.cc (hunk 1 above)
Comment (TaekyungHeo) [reply astra-sim/workload/HardwareResource.cc:38]: Having a CPU op is not mandatory from my understanding.

File: astra-sim/workload/Workload.cc (hunk 10 above)
Comment (TaekyungHeo) [reply astra-sim/workload/Workload.cc:178]: This function is only for compute operators.

File: astra-sim/workload/Workload.cc (hunk 2 above)
Comment (TaekyungHeo) [reply astra-sim/workload/Workload.cc]: Updated!

File: astra-sim/workload/Workload.cc (hunk 8 above)
Comment (TaekyungHeo) [reply astra-sim/workload/Workload.cc]: It has been replaced with issue_replay.

File: astra-sim/workload/Workload.cc (hunk 11 above)
Comment (TaekyungHeo) [reply astra-sim/workload/Workload.cc]: Updated.

File: astra-sim/workload/Workload.cc [hunk 16]
```diff
@@ -103,175 +102,201 @@ void Workload::issue_dep_free_nodes() {
...
-           << ",node->name=" << node->getChakraNode()->name() << endl;
-    }
-    issue_remote_mem(node);
-  } else if (node->getChakraNode()->node_type() == ChakraNodeType::COMP_NODE) {
-    if ((node->getChakraNode()->simulated_run_time() == 0) &&
-        (node->getChakraNode()->num_ops() == 0)) {
-      skip_invalid(node);
-    } else {
+  if (sys->replay_only) {
+      issue_replay(node);
```
Comment (TaekyungHeo) [inline astra-sim/workload/Workload.cc:112]: @beckmabd

File: astra-sim/workload/Workload.cc [hunk 17]
```diff
@@ -103,175 +102,201 @@ void Workload::issue_dep_free_nodes() {
...
   } else {
-    WorkloadLayerHandlerData* wlhd = new WorkloadLayerHandlerData;
-    wlhd->node_id = node->getChakraNode()->id();
-
-    sys->register_event(
-        this,
-        EventType::General,
-        wlhd,
-        node->getChakraNode()->simulated_run_time());
+    issue_replay(node);
```
Comment (TaekyungHeo) [inline astra-sim/workload/Workload.cc:195]: @beckmabd

Comment (beckmabd) [reply astra-sim/workload/Workload.cc:195]: OK, here is where calling the function issue_replay is a little confusing.  How about adding a comment stating "// advance this node forward the recorded "replayed" time specificed in the ET".

**Review by beckmabd — APPROVED** [review]
Can you just add one more comment.  Otherwise, this looks great!  Thanks for accommodating my requests.

File: astra-sim/workload/Workload.cc (hunk 17 above)
Comment (TaekyungHeo) [reply astra-sim/workload/Workload.cc:195]: Assigned to @JoongunPark.

Comment (TaekyungHeo) [reply astra-sim/workload/Workload.cc:197]: Resolved by @JoongunPark . @JoongunPark , please leave a comment here after you resolve a comment.

**Pull request merged**